pkg/kn/commands/plugin/plugin_list.go

```diff
@@ -0,0 +1,268 @@
+				fmt.Fprintf(o.ErrOut, "Unable read directory %q from your PATH: %v. Skipping...", dir, err)
+				continue
+			}
+
+			pluginErrors = append(pluginErrors, fmt.Errorf("error: unable to read directory %q in your PATH: %v", dir, err))
+			continue
+		}
+
+		for _, f := range files {
+			if f.IsDir() {
+				continue
+			}
+			if !hasValidPrefix(f.Name(), ValidPluginFilenamePrefixes) {
+				continue
+			}
+
+			if isFirstFile {
+				fmt.Fprintf(o.ErrOut, "The following compatible plugins are available:\n\n")
+				pluginsFound = true
+				isFirstFile = false
+			}
+
+			pluginPath := f.Name()
+			if !o.NameOnly {
+				pluginPath = filepath.Join(dir, pluginPath)
+			}
+
+			fmt.Fprintf(o.Out, "%s\n", pluginPath)
+			if errs := o.Verifier.Verify(filepath.Join(dir, f.Name())); len(errs) != 0 {
+				for _, err := range errs {
+					fmt.Fprintf(o.ErrOut, "  - %s\n", err)
+					pluginWarnings++
+				}
+			}
+		}
+	}
+
+	if !pluginsFound {
+		pluginErrors = append(pluginErrors, fmt.Errorf("error: unable to find any kubectl plugins in your PATH"))
+	}
+
+	if pluginWarnings > 0 {
+		if pluginWarnings == 1 {
+			pluginErrors = append(pluginErrors, fmt.Errorf("error: one plugin warning was found"))
+		} else {
+			pluginErrors = append(pluginErrors, fmt.Errorf("error: %v plugin warnings were found", pluginWarnings))
+		}
+	}
+	if len(pluginErrors) > 0 {
+		fmt.Fprintln(o.ErrOut)
+		errs := bytes.NewBuffer(nil)
+		for _, e := range pluginErrors {
+			fmt.Fprintln(errs, e)
+		}
+		return fmt.Errorf("%s", errs.String())
+	}
+
+	return nil
+}
+
+type PluginFlags struct {
+	NameOnly bool
+
+	Verifier    PathVerifier
+	PluginPaths []string
+
+	genericclioptions.IOStreams
+}
+
+func (p *PluginFlags) AddPluginFlags(command *cobra.Command) {
+	command.Flags().BoolVar(&p.NameOnly, "name-only", false, "If true, display only the binary name of each plugin, rather than its full path")
+}
+
+// pathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) []error
+}
+
+type CommandOverrideVerifier struct {
+	root        *cobra.Command
+	seenPlugins map[string]string
+}
+
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *CommandOverrideVerifier) Verify(path string) []error {
+	if v.root == nil {
+		return []error{fmt.Errorf("unable to verify path with nil root")}
+	}
+
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
+	if len(cmdPath) > 1 {
+		// the first argument is always "kubectl" for a plugin binary
+		cmdPath = cmdPath[1:]
+	}
+
+	errors := []error{}
+
+	if isExec, err := isExecutable(path); err == nil && !isExec {
+		errors = append(errors, fmt.Errorf("warning: %s identified as a kubectl plugin, but it is not executable", path))
```

Comment:
kubectl?
By default I'm not sure there's value in telling people about this condition. What if they really do need to have some random text file on disk that starts with `kn-` that just happens to be in their PATH... they're going to continually see this error message when it's not an error condition.   We shouldn't force them to rename it.

Reply:
Again see my global comment. By design.

Reply:
I think the design is wrong :-) 
Put it this way, I think a file in the PATH should only be considered a Kn plugin if:
- it's name starts with `kn-`
- it's executable bit is set

if a file in the PATH doesn't match this then it's not an error condition, it's the user's way of saying "skip it".